src/corelibs/U2Core/src/io/IOAdapterTextStream.cpp

```diff
@@ -0,0 +1,204 @@
+    : ioAdapter(_ioAdapter) {
+    ioDevice.reset(new IOAdapterDevice(ioAdapter));
+    stream.setDevice(ioDevice.data());
+
+    // Set the correct codec, otherwise QTextStream fallbacks to the current locale default (usually UTF-8).
+    if (codec != nullptr) {
+        stream.setCodec(codec);
+    }
+}
+
+GUrl IOAdapterReaderAndWriterBase::getURL() const {
+    return ioAdapter->getURL();
+}
+
+IOAdapterFactory *IOAdapterReaderAndWriterBase::getFactory() const {
+    return ioAdapter->getFactory();
+}
+
+///////////////////////////////////////////////
+////      IOAdapterReader
+///////////////////////////////////////////////
+IOAdapterReader::IOAdapterReader(IOAdapter *ioAdapter)
+    : IOAdapterReaderAndWriterBase(ioAdapter), unreadCharsBufferPos(0) {
+}
+
+IOAdapterReader::~IOAdapterReader() {
+    if (unreadCharsBufferPos < unreadCharsBuffer.length() && ioAdapter->isOpen()) {
+        // Roll IOAdapter back to the size of the data left in the unreadCharsBuffer.
+        // This is needed to correctly support streaming (reading multiple documents from the same file using IOAdapterReader)
+        int nCharsToReturn = unreadCharsBuffer.length() - unreadCharsBufferPos;
+        int nBytesToReturn = stream.codec()->fromUnicode(unreadCharsBuffer.right(nCharsToReturn)).length();
+        stream.seek(stream.pos() - nBytesToReturn);
+    }
+}
+
+int IOAdapterReader::read(U2OpStatus &os, QString &result, int maxLength, const QBitArray &terminators, IOAdapter::TerminatorHandling terminatorMode, bool *terminatorFound) {
+    CHECK_OP(os, 0);
+    result.clear();
+    textForUndo.clear();
+    bool isReadingTerminatorSequence = false;
+    while (!stream.atEnd() && result.length() != maxLength) {
+        QChar unicodeChar = get();
+        uchar latin1Char = unicodeChar.toLatin1();
+        bool isTerminatorChar = terminators.at(latin1Char);
+        if (isTerminatorChar) {
+            isReadingTerminatorSequence = true;
+            if (terminatorFound != nullptr) {
+                *terminatorFound = true;
+            }
+            if (terminatorMode == IOAdapter::Term_Exclude) {
+                unget();    // Push back the current terminator char and break.
+                break;
+            } else if (terminatorMode == IOAdapter::Term_Skip) {
+                continue;
+            }
+            // The terminator char will be included below as a normal one.
+        } else if (isReadingTerminatorSequence) {
+            unget();    // Push back the current non-terminator char and break.
+            break;
+        }
+        result.append(unicodeChar);
+    }
+
+    if (ioAdapter->hasError()) {
+        os.setError(ioAdapter->errorString());
+    }
+    return result.length();
+}
+
+QChar IOAdapterReader::get() {
+    QChar ch;
+    if (unreadCharsBuffer.isEmpty()) {
+        stream >> ch;
+    } else {
+        ch = unreadCharsBuffer[unreadCharsBufferPos];
```

Comment:
If `unreadCharsBuffer `doesn't have a char with index `unreadCharsBufferPos`? I offer to add a SAFE_POINT

Reply:
Added

Reply:
I have a second thought here: there is no "safe" way to return any character value from the SAFE_POINT here. If something is wrong here "crash" or "exception" is the best solution.
Returning of a non-meaningful character is wrong because it may lead to unpredictable behavior. And crash is better than any unpredictable behavior.

Reply:
Crash or unpredictable behavior is the worst thing we can have in this very often used class. This problem HAS to be resolved, and as far as this class is a part of the solution of this issue, this problem HAS to be resolved right here and right now. 
As far as any symbol could be met in file, we can't use one of them as an error symbol (I mean, symbol which is returned in case of the error). The best thing, I think, is to throw the exception and catch it one level below on the call stack.

Reply:
Why do you think there is a problem to solve here? Both variables are managed internally only and they are never out of sync.

We have tests that prove that the variables state is always correct and will have even more after all formats are migrated. 
Do you have any scenario where there is a chance of error here? If yes, lets add the test?

Because otherwise you ask me to add a boilerplate code for the impossible situation. This code will spoil the readability and will never work: the situation is impossible. 

The extra checks will be valid if the state could be affected by user, but here it is purely internal state.
Here is an example of the similar QT code: https://code.woboq.org/qt5/qtbase/src/corelib/io/qbuffer.cpp.html

Reply:
Finally I think I found an option that will suite both of us:
 I added SAFE_POINT with os.setError() inside. Reporting errors via OpStatus is the way this class works for all other cases.

Reply:
Yes, I like this solution.